seacrowd/sea_datasets/melayu_brunei/melayu_brunei.py

```diff
@@ -0,0 +1,203 @@
+}
+
+_SUPPORTED_TASKS = [Tasks.SELF_SUPERVISED_PRETRAINING]
+
+
+_SOURCE_VERSION = "1.0.0"
+
+_SEACROWD_VERSION = "1.0.0"
+
+
+# TODO: Name the dataset class to match the script name using CamelCase instead of snake_case
+class MelayuBruneiDataset(datasets.GeneratorBasedBuilder):
+    """This article gives two texts of Brunei Malay (ISO 639-3: kxd) collected in the town of
+    Weston in Sabah State of Malaysia. The texts exhibit linguistic features that are similar to those of
+    Brunei Malay spoken in Brunei Darussalam; it has a vowel inventory of only three vowels /a, i, u/,
+    use of the pronoun kitani for the first person plural inclusive and the use of the base-stem transitive
+    form in patientive voice clauses. One of the texts tells a folk story about Batang Dayang and other text
+    includes two episodes: Javanese runaways arriving in Weston and the origin of the name Bukau, a town near Weston.
+"""
+
+    SOURCE_VERSION = datasets.Version(_SOURCE_VERSION)
+    SEACROWD_VERSION = datasets.Version(_SEACROWD_VERSION)
+
+    SEACROWD_SCHEMA_NAME = TASK_TO_SCHEMA[_SUPPORTED_TASKS[0]].lower()
+
+    BUILDER_CONFIGS = [
+        SEACrowdConfig(
+            name=f"{_DATASETNAME}_source",
+            version=SOURCE_VERSION,
+            description=f"{_DATASETNAME} source schema",
+            schema="source",
+            subset_id=f"{_DATASETNAME}",
+        ),
+        SEACrowdConfig(
+            name=f"{_DATASETNAME}_seacrowd_{SEACROWD_SCHEMA_NAME}",
+            version=SEACROWD_VERSION,
+            description=f"{_DATASETNAME} SEACrowd schema",
+            schema=f"seacrowd_{SEACROWD_SCHEMA_NAME}",
+            subset_id=f"{_DATASETNAME}",
+        ),
+    ]
+
+    DEFAULT_CONFIG_NAME = f"{_DATASETNAME}_source"
+
+    def _info(self) -> datasets.DatasetInfo:
+
+        if self.config.schema == "source":
+            features = datasets.Features(
+                {
+                    "id": datasets.Value("string"),
+                    "text": datasets.Value("string"),
+                }
+            )
+
+        elif self.config.schema == f"seacrowd_{self.SEACROWD_SCHEMA_NAME}":
+            features = schemas.self_supervised_pretraining.features
+
+        return datasets.DatasetInfo(
+            description=_DESCRIPTION,
+            features=features,
+            homepage=_HOMEPAGE,
+            license=_LICENSE,
+            citation=_CITATION,
+        )
+
+    def _split_generators(self, dl_manager: datasets.DownloadManager) -> List[datasets.SplitGenerator]:
+        """Returns SplitGenerators."""
+        urls = [_URLS[key] for key in _URLS.keys()]
+        data_path = dl_manager.download_and_extract(urls)
+
+
+        return [
+            datasets.SplitGenerator(
+                name=datasets.Split.TRAIN,
+                gen_kwargs={
+                    "filepath": data_path[0],
+                    "split": "train",
+                    "other_path": data_path[1:]
```

Comment:
Is there a particular reason why we need to separate `filepath` and the `other_path`? The `gen_kwargs` is arbitrary, so you can use `filepaths` instead and pass the full list.

Reply:
This is because there are many URLs. If we feed all URLs directly to the `gen_kwargs`, we will receive an error showing that the string length exceeds the limit.

Reply:
Hi @Alex-HaochenLi , I think I'm still a bit confused. I'm thinking something like this:

```python
...
gen_kwargs={
    "filepaths": data_path  # this is a list of string
}
...
```

And then in `_generate_examples`, we can just remove this line:

```
# Remove this line
filepaths = [filepath] + other_path
```

It seems to just reconstruct the list in the earlier function, I'm not sure why this step is needed. 
Let me know if I missed anything!

Reply:
Hello @ljvmiranda921. My initial idea is to implement what you described. However, I found that this implementation causes an error of filename too long. The detail of the error message are shown below. In short, this is because there are too many file URLs. 

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/haochen/anaconda3/envs/env-seacrowd/lib/python3.10/site-packages/datasets/builder.py", line 1724, in _prepare_split_single
    for key, record in generator:
  File "/home/haochen/.cache/huggingface/modules/datasets_modules/datasets/melayu_brunei/e95f90a4b437b9997afd7357a74c2c1f92d7f13bcbb147d3959dae1012f9dbbf/melayu_brunei.py", line 189, in _generate_examples
    with open(filepath, "r") as f:
  File "/home/haochen/anaconda3/envs/env-seacrowd/lib/python3.10/site-packages/datasets/streaming.py", line 75, in wrapper
    return function(*args, download_config=download_config, **kwargs)
  File "/home/haochen/anaconda3/envs/env-seacrowd/lib/python3.10/site-packages/datasets/download/streaming_download_manager.py", line 501, in xopen
    return open(main_hop, mode, *args, **kwargs)
OSError: [Errno 36] File name too long: '[/home/haochen/.cache/huggingface/datasets/downloads/a2a391e70fde8e8baf41c1289627d024b3ba0104b6c34c6694c5c95a795c2428 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-1-01.txt), /home/haochen/.cache/huggingface/datasets/downloads/42906723290bbd703b704e74239f5fdcd393cb92fefbbef4843d01504343f202 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-1-02.txt), /home/haochen/.cache/huggingface/datasets/downloads/e82b299a45dd3c9186023c99397ad59a5b53034df5c37ac795fb7be1ec5f9998 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-1-03.txt), /home/haochen/.cache/huggingface/datasets/downloads/4430f836ac64376954ae0da065b66ac19911d4a6688949416536cc95ba8d2e32 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-1-04.txt), /home/haochen/.cache/huggingface/datasets/downloads/b356f637535eba45f0baca1c701f189a19ed7c5bb7a11cd3b3abbbfd95e179e1 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-1-05.txt), /home/haochen/.cache/huggingface/datasets/downloads/4a8d42af0a6bcf6a5966ba079dd7d56c2c13be31d05cbebc51c6f2dbecfca5a2 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-1-06.txt), /home/haochen/.cache/huggingface/datasets/downloads/a11985567a50eee5f7369936d45189b065243115cb93d05e4e43bbe5d588e561 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-1-07.txt), /home/haochen/.cache/huggingface/datasets/downloads/f3dced89e4c12bf72cf880bcd2f6a8d56e8bca5fa87b58af599749f48a7d49fd (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-1-08.txt), /home/haochen/.cache/huggingface/datasets/downloads/ba7a49b8b314298aa33ede17220af749bade06685373b0c76163e134be465058 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-2-01.txt), /home/haochen/.cache/huggingface/datasets/downloads/e7557d3693858825d5637e1453a075b36745687580769ee5f8696226015954c3 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-2-02.txt), /home/haochen/.cache/huggingface/datasets/downloads/bfd947989007a8c0af60c018687d672c18bd2e2058b6aa55211848836225fdcc (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-2-03.txt), /home/haochen/.cache/huggingface/datasets/downloads/ea36c00762d2c77b8c790ab5522f72a8b2e18e2ff68cb929ff6924c6a5fc4543 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-2-04.txt), /home/haochen/.cache/huggingface/datasets/downloads/b23a8dd43a5699fdb492f9b8e59c1bd4c1e690b6dacd82ab5f168547ad7b574a (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-2-05.txt), /home/haochen/.cache/huggingface/datasets/downloads/ed8070158df615c6756a7d20f33fdbeca30e1927b31c55b5e99dd8e6b881dcb4 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-2-06.txt), /home/haochen/.cache/huggingface/datasets/downloads/09d2dd16b75fe6fa6f160ef9991e4fafdc7fc22182b5333ad2e104c064d7f1f1 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-2-07.txt), /home/haochen/.cache/huggingface/datasets/downloads/6c51776073bbed93cd3ef49ab4dccbdc99eb7bc2246a4ca0f5feaf6fac913f00 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-2-08.txt), /home/haochen/.cache/huggingface/datasets/downloads/9dea087a30ef6f3fe22a104327a9d5c27709414845fc531aa4c54bd4d48bb172 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-2-09.txt), /home/haochen/.cache/huggingface/datasets/downloads/154f6fea130f1941a671732d08b40713b43a420aa03aa14a6a1123a22acb5557 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-2-10.txt), /home/haochen/.cache/huggingface/datasets/downloads/f7b2d72255783f554876ac47751aa5fa32c06463e58a40ebbb797c4ed0020c9d (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-2-11.txt), /home/haochen/.cache/huggingface/datasets/downloads/2a80acb5c8cb2b3a5df6d8a94a422730a959e0d39a2f0b96efabcd1ab65c7d64 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-2-12.txt), /home/haochen/.cache/huggingface/datasets/downloads/f0c2cb32f7e01df873a0742b6f26104391f9dc171e556a1c27536cbcfc930952 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-2-13.txt), /home/haochen/.cache/huggingface/datasets/downloads/a32c3c5f9ba9c385b86f8425269e0de548642251f5527c235aa0681f624ef1cf (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-2-14.txt), /home/haochen/.cache/huggingface/datasets/downloads/7cc8409e3c1bc277eaaede0c0a5e9c3b9479191b9306a509b8c9b2d025c03ca2 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-2-15.txt), /home/haochen/.cache/huggingface/datasets/downloads/c82b8e996a089a0b937983b72689bdae80fa79fc01bd3f5a30b63c9c59dad615 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-2-16.txt), /home/haochen/.cache/huggingface/datasets/downloads/05a800b43d5a5b7f183fa4903c96f8651ddd3543bcb27f349d0b3ad70167d2aa (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-3-01.txt), /home/haochen/.cache/huggingface/datasets/downloads/d4b350893cc0dd42809d923fc5b58c9d69d4c7ab666c90f93712f3c4dc317b74 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-3-02.txt), /home/haochen/.cache/huggingface/datasets/downloads/428b41649b6e3d42f9de8241528b30f7ed519f0705d4ab151466040c0823000a (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-3-03.txt), /home/haochen/.cache/huggingface/datasets/downloads/871989d6c689c9f6f3c7658eec1637d45811e6086468d2193d7180b5869d138c (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-3-06.txt), /home/haochen/.cache/huggingface/datasets/downloads/22ea5223f255d7f119cfc9728e305d9f515233b13122ea9bf9b481de0a798d8d (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-3-07.txt), /home/haochen/.cache/huggingface/datasets/downloads/1ad8ce4b91dfbd434f85bb884dc3eaa381766f3a6373cb8200fb68c6434b09b8 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-3-08.txt), /home/haochen/.cache/huggingface/datasets/downloads/4daad7fea3e439180203117ac97407178d91c9e1658a4b267706289cf44fff9e (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-3-09.txt), /home/haochen/.cache/huggingface/datasets/downloads/4a6bc4c9e3eaebc07ee396f72de76d355bf07f82e91b914cf97402b35105c0c2 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-3-10.txt), /home/haochen/.cache/huggingface/datasets/downloads/9b7ec8972d634f5f038202e65197c4359fcf43e8414abdcd3552a64adc0c72a9 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-4-00.txt), /home/haochen/.cache/huggingface/datasets/downloads/b58ed8e4940057c3405cbb2d0d0b8be7a3dc10036863e8787aa9b30941e8710f (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-4-01.txt), /home/haochen/.cache/huggingface/datasets/downloads/c7b0f4ea8e97a3c6d956bc6ea7054b70bc93d520c54cf04fc45ec13834ff357e (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-4-02.txt), /home/haochen/.cache/huggingface/datasets/downloads/231f6f977dc0dc3940934277a39ea6ecc62fa37922c077dfc477203b68217125 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-4-03.txt), /home/haochen/.cache/huggingface/datasets/downloads/28cb3b0907b1037beef5bd2c4194f525eae3b9df582af13e0ede7ca2be277e06 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-4-04.txt), /home/haochen/.cache/huggingface/datasets/downloads/f6a60171bace404a963dc253593c8a22293fc1582edffbc0f70d31e27a381d08 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-4-05.txt), /home/haochen/.cache/huggingface/datasets/downloads/e5b415b80cd7a5ef13738bfa04e76e2cb1503b4d44cc58ebaf2708ae9d4bb2a0 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-4-06.txt), /home/haochen/.cache/huggingface/datasets/downloads/4b0c28899305b7c493abc177c9f7701fe512e4eb0952c89e34cc311ec54ffa24 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-4-07.txt), /home/haochen/.cache/huggingface/datasets/downloads/781b3e48a3f4af8ab721888eb9bea4fa9b6bd6ac107dc57e62b77c8bda926172 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-4-08.txt), /home/haochen/.cache/huggingface/datasets/downloads/5a7b27c5158a587889ce32e4d7222e89e17afbd6dad3d1f7745053a208670db4 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-4-09.txt), /home/haochen/.cache/huggingface/datasets/downloads/6029f83ae0334c436a6d6d9c443a886ed6460c5d8cee925c54a488639f347dc1 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-5-01.txt), /home/haochen/.cache/huggingface/datasets/downloads/82e17960dd79e77a0b96d4ce15b2795daeb1b042ff499c170dd2f1f8a2397940 (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-5-02.txt), /home/haochen/.cache/huggingface/datasets/downloads/ea94c93a6d7005c9b46d19e2167d46b4f6497ec07f3d49e02e7198537837a99a (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor2-5-03.txt), /home/haochen/.cache/huggingface/datasets/downloads/e512ad5c9b64be3237dc2f47b4e3d3895b9bf98c609dfd6197e4c3ddeaac465b (origin=https:/raw.githubusercontent.com/matbahasa/Melayu_Brunei/master/Folklor/Folklor3-0-01.txt)]'

Reply:
Ah that's an interesting error. Ok thanks for explaining!